### ISSUE TITLE
Services without selectors cannot forward to other services as documented

### ISSUE THREAD

**Issue by robhaswell** [kubernetes/kubernetes#40870]
**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:57:05Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:

Google Container Engine, 1.5.2

**What happened**:

I created a selectorless service and configured the endpoint to point at the ClusterIP of another service in a different namespace, as described in the documentation: 
https://kubernetes.io/docs/user-guide/services/#services-without-selectors

This is supported behaviour ("You want to point your service to a service in another Namespace or on another cluster").

Any requests on the External IP of the selectorless service hang.

**What you expected to happen**:

Requests to be sent to the target service.

**How to reproduce it** (as minimally and precisely as possible):

Start this stack:

```yaml
---
kind: Namespace
apiVersion: v1
metadata:
  name: reproducer-alpha
---
kind: Pod
apiVersion: v1
metadata:
  name: alpha
  namespace: reproducer-alpha
  labels:
    service: alpha
spec:
  containers:
    - name: alpha
      image: robhaswell/hello-env
      env:
        - name: HELLO_MSG
          value: I am ALPHA
      ports:
        - containerPort: 8080
---
kind: Service
apiVersion: v1
metadata:
  name: alpha
  namespace: reproducer-alpha
spec:
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
  selector:
    service: alpha
  clusterIP: 10.3.250.250
---
kind: Service
apiVersion: v1
metadata:
  name: frontend
  namespace: default
spec:
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
  type: LoadBalancer
---
kind: Endpoints
apiVersion: v1
metadata:
  name: frontend
  namespace: default
subsets:
  - addresses:
    - ip: 10.3.250.250
    ports:
    - port: 8080
```

Wait for the `frontend` service to be allocated an IP, then try to hit it with curl on port 8080: `curl http://EXTERNAL-IP:8080`. There is no output. I would expect the output to be 'I am ALPHA'.

**Anything else we need to know**:

**Comment by taimir** [comment]
Hi there, this seems interesting:

AFAICT, the problem seems to be a mismatch between the docs and what the `iptables proxier` in `kube-proxy` actually can do. In `k8s`, the otherwise purely virtual `Services` are realized by smart DNAT-ing on each node in the cluster (done by `kube-proxy`). The DNAT rules are more or less the core functionality, and are defined in the NAT table of `iptables`, on each node, for each existing endpoint of each service in the cluster.

From what I can see, the problem in your case is that the DNAT in `iptables` is effectively done only once (all the `iptables` chains are traversed just once per packet, see [here](http://inai.de/images/nf-packet-flow.png)), so that the request to the `frontend` service is effectively resolved to a request to `10.3.250.250:8080` after it comes out of `netfilter` (i.e. after all `iptables` chains are applied). But, `10.3.250.250` is not a real IP (it's a virtual service IP), and thus your requests fails as it has no real destination.

One way to deal with this is to make `kube-proxy` **recursively** resolve the DNAT rules in case there are chained services as in your case, and then sync them with `iptables` (the sync is done in real time anyways). [Here](https://github.com/taimir/kubernetes/commit/563118116f27d2c0f59073d264d07cadb698b5ed) is a sketch commit of how this can be "fixed" in the code (it works in my dev setup).

It would be good to hear a second opinion on whether this use case should indeed be supported. In case you are interested, I can open a PR with the fix and try to merge it.

**Comment by taimir** [comment]
FYI, the service-to-service forwarding chain didn't work in my dev cluster even if both services are in the same namespace (as the problem described above seems to be conceptual). The services being in different namespaces is a whole separate issue on it's own.

**Comment by robhaswell** [comment]
Yes, I apologise, I noticed that this functionality doesn't work even in the same namespace while constructing this issue, but did a bad job of removing the references to namespacing.

**Comment by kargakis** [comment]
@kubernetes/sig-network-misc

**Comment by thockin** [comment]
It's not surprising to me that this doesn't work.  I don't think this
particular usage pattern was  tested or even considered really.

It's not invalid, but it's a little weird.  Can you use ExternalName
instead, to achieve the desired effect?

**Comment by robhaswell** [comment]
> I don't think this particular usage pattern was  tested or even considered really.

@thockin I'm confused, my usage pattern matches exactly what is [documented as a use-case for this functionality](https://kubernetes.io/docs/user-guide/services/#services-without-selectors):

> You want to point your service to a service in another Namespace or on another cluster.

Please could you describe how my usage pattern is different?

I might be able to use `ExternalName`, it depends on the TTL, but in my experience using DNS for traffic routing is a bad idea.

**Comment by kewne** [comment]
> It's not invalid, but it's a little weird.  Can you use ExternalName
instead, to achieve the desired effect?

I got this working with ExternalName, just put the FQDN in the externalName property.
I'm running on an on-prem cluster running 
```
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+coreos.0", GitCommit:"3ed7d0f453a5517245d32a9c57c39b946e578821", GitTreeState:"clean", BuildDate:"2017-01-13T00:23:19Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

> I might be able to use ExternalName, it depends on the TTL, but in my experience using DNS for traffic routing is a bad idea.

Actually, according to the description in the docs, ExternalName creates a DNS CNAME which, according to wikipedia (my DNS knowledge isn't that great), is:
> used to specify that a domain name is an alias for another domain, which is the "canonical" domain.

This seems to be semantically consistent with what we're doing.

**Comment by caseydavenport** [comment]
> It would be good to hear a second opinion on whether this use case should indeed be supported

> I got this working with ExternalName

Yeah, IMO we can probably turn this into a documentation issue.  Sounds like the "feature" is supported through ExternalName, but the docs don't make it clear how to achieve this.

**Comment by yelled3** [comment]
/reopen
@robhaswell did you solve this issue?

currently the documentation hasn't changed, BTW.

see:
https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors

**Comment by robhaswell** [comment]
@yelled3 no, I never managed to get this working as documented. I gave up on this issue as soon as @thockin said "I don't think this particular usage pattern was  tested or even considered really" when my exact use-case was listed as an example in the documentation ("You want to point your service to a service in another Namespace or on another cluster").

I've encountered enough _wontfix_ attitudes on open source projects to know that pursuing this would exceed the amount of time I have available to finish this work.

In the end I used `haxproxy` in TCP mode and configured it with environment variables.
